### DISCUSSION TITLE
URL slugs: How to handle duplicates?

### DISCUSSION THREAD

**Discussion by nguillaumin** [atari-legend/atari-legend#211] [General] [answered]
I'm looking to implement URL slugs, inspired by what Lemon64 did (I actually had it on my list since the beginning of the rewrite!). So a game URL would contain the game name, rather than the ID. For example https://www.atarilegend.com/games/978 would become https://www.atarilegend.com/games/dungeon-master or https://www.atarilegend.com/games/3441 would become https://www.atarilegend.com/games/zork-2-the-wizard-of-frobozz  . It's a bit nicer and it should give us a SEO boost.

The slugs will be automatically generated but can be manually tweaked in CPANEL if needed. I'm looking at generating slugs for all the games we already have but one issue is that we have a lot of games that have the same name. I think it's mostly public domain games. For example:
- Othello: https://www.atarilegend.com/games/search?title=othello
- Tetris: https://www.atarilegend.com/games/search?title=tetris
- Joust: https://www.atarilegend.com/games/search?title=joust

I tried a heuristic that if I find a duplicate I append the developer name, and then one of the individual name. So for example [Tetris by AG Soft](http://localhost:8080/games/7301) would be `/tetris-ag-soft`. Othello by [Petri Sinisalo](http://localhost:8080/games/5897) will be `/othello-petri-sinisalo`, etc. But despite this I still get duplicates as for a lot of these there is no developer or individual.

I was wondering if anyone had a good idea about how to handle that? The only one I have is to append the ID, so `/othello-6921` for example. Then later one we can see if we can find an individual name, enter it in the DB, and update the slug (but this would be manual correction steps). I can see in some screenshots that there are individual names for example...

Any better ideas?

**Comment by stgraveyard** [comment]
Hi Nico,

No, I don't have a better idea. Only if no dev or pub or ind exist, use the id. Is thay possible?

Cheers, 
Maarten

**Reply by nguillaumin** [reply]
Yes, I think it's basically what I propose. In any case the slug will be editable in CPANEL, so if a DB user finds a better one they can change it afterwards.

**Comment by Brume-AL** [comment]
Hello,

Good idea, Nico. As you suggest it, I guess adding the ID to the name of the game is the best way to deal with the duplicated titles.
I see a (small) issue with this kind of URL: if the title of the game has a number, it may be confusing for Google.
Cobra 2 will have this kind of URL, right?
https://www.atarilegend.com/games/cobra-2-850

So maybe could we place the ID somewhere else? Something like that, maybe?
https://www.atarilegend.com/games/850/cobra-2

Also, if we want to increase our SEO position, we could also give some real name to the pictures. Google doesn't like images named 659.png, 660.png, 661.png, etc. Is there an easy way to rename them as gods01.png, gods02.png, etc.?
Last but not least, we could also add a description to the images in the metadatas, too (Gods by Bitmap Brothers, for instance).

Let me know what you think.

Thanks,

Bruno

**Reply by nguillaumin** [reply]
Thanks for the input!

> https://www.atarilegend.com/games/850/cobra-2

I would like to avoid keeping the ID in the URL at all if possible, so I'd rather not put it in the URL path as a fixed section.

Maybe it's not too much of an issue because that would only appear if there's a duplicate. For example there's only one "Cobra 2" so the slug will be `/cobra-2`. Probably there are not many cases which are a duplicate + end with a number?
Possibly we could add an `-id-<id>` suffix to clarify, like `/othello-id-1234`.

Keep in mind the slug will be editable in CPANEL so if we find a better one we can change it easily (e.g. `/othello-pd-st-format-12` for example if it come from a cover disk).

> give some real name to the pictures
> add a description to the images

Great ideas! I think once I have the basic game slug going, I can implement that easily.

**Comment by nguillaumin** [comment, marked as answer]
Hi, I have deployed the slugs on the dev site: https://dev.atarilegend.com/ . Please have a look and let me know what you think.

For the duplicates I added the suffix `-id-1234` with the ID, for example: https://dev.atarilegend.com/games/joust-id-5502 . The slug is editable in CPANEL:

![slug](https://user-images.githubusercontent.com/762512/218315249-154f4f26-6f76-4c09-9512-6699e0e21cee.png)

So for example this one could be changed to `joust-softrunner`.

I have **not** implemented Brume's idea yet, I wanted to get the basics going first.

**Reply by nguillaumin** [reply]
I also changed the screenshots URL to include the game name: https://dev.atarilegend.com/games/sundog-frozen-legacy/screenshot-621.png  . Unfortunately I could not make it as good as we wanted (`/games/sundog-frozen-legacy01.jpg`) without a lot more work.

**Reply by nguillaumin** [reply]
This is now deployed!
